include/seqan3/search/fm_index/fm_index.hpp

```diff
@@ -525,3 +581,62 @@ fm_index(text_t &&) -> fm_index<range_innermost_value_t<text_t>, text_layout{ran
 
 //!\}
 } // namespace seqan3
+
+namespace seqan3::detail
+{
+
+/*!\brief An FM Index specialisation that handles reversing the given text.
+ * \implements seqan3::fm_index_specialisation
+ * \tparam alphabet_t        The alphabet type; must model seqan3::semialphabet.
+ * \tparam text_layout_mode_ Indicates whether this index works on a text collection or a single text.
+ *                           See seqan3::text_layout.
+ * \tparam sdsl_index_type_  The type of the underlying SDSL index, must model seqan3::sdsl_index.
+ * \implements seqan3::cerealisable
+ *
+ * \details
+ *
+ *  This FM Index reverses the given text before constructing the seqan3::fm_index.
+ *  This type is used by the seqan3::bi_fm_index.
+ */
+template <semialphabet alphabet_t,
+        text_layout text_layout_mode_,
+        detail::sdsl_index sdsl_index_type_ = default_sdsl_index_type>
+class reverse_fm_index : public fm_index<alphabet_t, text_layout_mode_, sdsl_index_type_>
+{
+private:
+    //!\copydoc fm_index::construct()
+    template <std::ranges::range text_t>
+    //!\cond
+        requires (text_layout_mode_ == text_layout::collection)
+    //!\endcond
+    void construct_(text_t && text)
+    {
+        auto reverse_text = text | views::deep{std::views::reverse} | std::views::reverse;
+
+        this->construct(reverse_text, true);
+    }
+
+    //!\copydoc fm_index::construct()
+    template <std::ranges::range text_t>
+    //!\cond
+        requires (text_layout_mode_ == text_layout::single)
+    //!\endcond
+    void construct_(text_t && text)
```

Comment:
Can't these overloads be avoided with a simple `if constexpr` inside the function?